### PR TITLE
Test branch

### PR DESCRIPTION
**Pull request by Josphat205**
### **In this Branch**

**Added the following validations:**
**For the User model:**

-  Name must not be blank.

- PostsCounter must be an integer greater than or equal to zero.

**For the Post model:**

- Title must not be blank.

- Title must not exceed 250 characters.

- CommentsCounter must be an integer greater than or equal to zero.

- LikesCounter must be an integer greater than or equal to zero.

- Added unit specs for all of your models' methods and validations.

### CODE REVIEW

File: spec/models/user_spec.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  it 'posts_counter should be >= 0' do
+    @user.posts_counter = -2
+    expect(@user).to_not be_valid
+  end
+
+  it 'posts_counter should be numeric' do
+    @user.posts_counter = 'One'
+    expect(@user).to_not be_valid
+  end
+end
```
Comment (DammyShittu) [inline spec/models/user_spec.rb]: - [x] Methods in the models should be tested as well to cover all possible scenarios. Kindly add test for the `fetch_recent_posts` method in your User model file.

File: spec/models/post_spec.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+  it 'should not accept more than 250 character' do
+    @post.title = '
+    Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Aenean commodo ligula
+    eget dolor. Aenean massa. Cum sociis natoque penatibus et magnis dis parturient
+    montes, nascetur ridiculus mus. Donec quam felis, ultricies nec, pellentesque
+    eu, pretium quis,'
+    expect(@post).to_not be_valid
+  end
+end
```
Comment (DammyShittu) [inline spec/models/post_spec.rb]: - [x] Kindly add tests for the `fetch_recent_comments` method in your Post model.

File: spec/models/comment_spec.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
+require 'rails_helper'
+
+RSpec.describe Comment, type: :model do
```
Comment (DammyShittu) [inline spec/models/comment_spec.rb:3]: - [x] You have done well adding tests for the User, Post, and Comment models, but all models should be tested. Kindly add the missing tests for the Like model also.

**Review by DammyShittu — CHANGES_REQUESTED** [review]
Hi @Josphat205,

## Highlights
✔️ Descriptive PR
✔️ Passing linters
✔️ Good branch name

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/models/user_spec.rb (hunk 1 above)
Comment (Josphat205) [reply spec/models/user_spec.rb]: Added

File: spec/models/post_spec.rb (hunk 2 above)
Comment (Josphat205) [reply spec/models/post_spec.rb]: Added

File: spec/models/comment_spec.rb (hunk 3 above)
Comment (Josphat205) [reply spec/models/comment_spec.rb:3]: Added

File: spec/models/user_spec.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    @user.save
+    @post = Post.create(
+      title: 'A new Post',
+      text: 'Web Developer',
+      author_id: @user.id,
+      comments_counter: 0,
+      likes_counter: 0
+    )
+    expect(@user.fetch_recent_posts.count).to eq(1)
+  end
```
Comment (BenMukebo) [inline spec/models/user_spec.rb]: - [x]  Great job adding the fetch recent posts test however it does not pass yet. Kindly fix this 💯 👍🏻 

 image

File: spec/models/comment_spec.rb [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+      likes_counter: 0
+    )
+  end
+
+  it '@users comments_counter should increase by one' do
+    @comment = Comment.create(text: 'Hello World!', author_id: @user.id, post_id: @post.id)
+    @comment = Comment.create(text: 'Welcome World!!', author_id: @user.id, post_id: @post.id)
+    expect(@comment.author_id).to_not eq(30)
+  end
+end
```
Comment (BenMukebo) [inline spec/models/comment_spec.rb]: - [OPTIONAL] The same here. So far great job 👏🏻 However,  you are missing a test to test the `update_counter` method of your Comment model, please add this ✍🏻

```ruby
  def update_counter
    post.increment!(:comments_counter)
  end
```

File: spec/models/like_spec.rb [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+      comments_counter: 0,
+      likes_counter: 0
+    )
+  end
+
+  it '@users comments_counter should increase by one' do
+    @comment = Comment.create(text: 'Hello World!', author_id: @user.id, post_id: @post.id)
+    @comment = Comment.create(text: 'Welcome World!!', author_id: @user.id, post_id: @post.id)
+    expect(@comment.author_id).to_not eq(30)
+  end
```
Comment (BenMukebo) [inline spec/models/like_spec.rb:24]: - [ ] The same here. As requested by the previous reviewer, Kindly add the missing tests for the Like model. You are missing a test to test the `update_likes_counter` method of your Like model, please add this ✍🏻

```ruby
  def update_likes_counter
    post.increment!(:likes_counter)
  end
```
- Note you need to test the method, actually, you're checking the author_id on line (23)

**Review by BenMukebo — CHANGES_REQUESTED** [review]
Hi @Josphat205, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

## To Highlight :clap: 
- Gitflow is followed. :heavy_check_mark:
- PR has a good title and summary. ✔️


 ## Required Changes ♻️
_Check the comments under the review._


You have submitted a project, that Linters check are not running in the GitHub/action
## Optional suggestions

- [OPTIONAL] I also recommend that you delete the Note: 👇🏻 in your License section as these are instructions on how to complete the License section.

image

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏 💻  
Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please Tag me on (@BenMukebo) when you comment so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/models/post_spec.rb [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    eu, pretium quis,'
+    expect(@post).to_not be_valid
+  end
+  it 'fetch_recent_comments' do
+    @user.save
+    @post.save
+    @comment = Comment.create(text: 'Hello World!', author_id: @user.id, post_id: @post.id)
+    @comment = Comment.create(text: 'Welcome World!!', author_id: @user.id, post_id: @post.id)
+    expect(@post.fetch_recent_comments.count).to eq(2)
+  end
```
Comment (BenMukebo) [inline spec/models/post_spec.rb:54]: - [OPTIONAL] Great job so far 👏🏻 However, all custom methods in the models must have a corresponding unit test. Actually, you are missing a test to test the `update_posts_counter` method of your Post model, please add this ✍🏻

```ruby
  def update_posts_counter
     author.increment!(:posts_counter)
  end
```

**Review by codepantha — APPROVED** [review]
# STATUS: APPROVED 🟢 ✔️ 

## Hi @Josphat205 👋, 

Well done for implementing the required changes 👏 
Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**